pkg/registry/apis/dashboard/authorizer.go

```diff
@@ -0,0 +1,95 @@
+package dashboard
+
+import (
+	"context"
+
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+
+	"github.com/grafana/grafana/pkg/apis/dashboard/v0alpha1"
+	"github.com/grafana/grafana/pkg/infra/appcontext"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/grafana-apiserver/endpoints/request"
+	"github.com/grafana/grafana/pkg/services/guardian"
+)
+
+func (b *DashboardsAPIBuilder) GetAuthorizer() authorizer.Authorizer {
+	return authorizer.AuthorizerFunc(
+		func(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+			if !attr.IsResourceRequest() {
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+
+			user, err := appcontext.User(ctx)
+			if err != nil {
+				return authorizer.DecisionDeny, "", err
+			}
+
+			if attr.GetName() == "" {
+				// Discourage use of the "list" command for non super admin users
+				if attr.GetVerb() == "list" && attr.GetResource() == v0alpha1.DashboardResourceInfo.GroupResource().Resource {
+					if !user.IsGrafanaAdmin {
+						return authorizer.DecisionDeny, "list summary objects (or connect as GrafanaAdmin)", err
+					}
+				}
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+
+			ns := attr.GetNamespace()
+			if ns == "" {
+				return authorizer.DecisionDeny, "expected namespace", nil
+			}
+
+			info, err := request.ParseNamespace(attr.GetNamespace())
+			if err != nil {
+				return authorizer.DecisionDeny, "error reading org from namespace", err
+			}
+
+			// expensive path to lookup permissions for a the single dashboard
+			dto, err := b.dashboardService.GetDashboard(ctx, &dashboards.GetDashboardQuery{
+				UID:   attr.GetName(),
+				OrgID: info.OrgID,
+			})
+			if err != nil {
+				return authorizer.DecisionDeny, "error loading dashboard", err
+			}
+
+			ok := false
+			guardian, err := guardian.NewByDashboard(ctx, dto, info.OrgID, user)
```

Comment:
The guardian appears to only use IsFolder, UID & ID from dto, so we could maybe use a more lightweight query to get that info.

Reply:
Looking deeper, ID is only used to build 2 debug log entries, so we could skip that altogether and just set IsFolder and UID.